src/d3a/models/area/redis_external_market_connection.py

```diff
@@ -41,15 +42,43 @@ def market_stats_callback(self, payload):
         payload_data = json.loads(payload["data"])
         ret_val = {"status": "ready",
                    "command": "market_stats",
-                   "market_stats":
-                       self.area.stats.get_market_stats(payload_data["market_slots"])}
+                   "market_fee_const":
+                       self.area.stats.get_market_stats(payload_data["market_slots"]),
+                   "transaction_id": payload_data.get("transaction_id", None)}
         self.publish(market_stats_response_channel, json.dumps(ret_val))
 
-    def grid_fees_callback(self, payload):
+    def set_grid_fees_callback(self, payload):
         market_stats_response_channel = f"{self.channel_prefix}/response/grid_fees"
         payload_data = json.loads(payload["data"])
         self.area.transfer_fee_const = payload_data["fee"]
         self.publish(market_stats_response_channel, json.dumps({
             "status": "ready", "command": "grid_fees",
-            "fee": str(self.area.transfer_fee_const)})
+            "market_fee_const": str(self.area.transfer_fee_const),
+            "transaction_id": payload_data.get("transaction_id", None)})
          )
+
+    def dso_market_stats_callback(self, payload):
+        market_stats_response_channel = f"{self.channel_prefix}/response/dso_market_stats"
+        payload_data = json.loads(payload["data"])
+        ret_val = {"status": "ready",
+                   "command": "dso_market_stats",
+                   "market_stats":
+                       self.area.stats.get_market_stats(payload_data["market_slots"]),
+                   "market_fee_const": str(self.area.transfer_fee_const),
+                   "transaction_id": payload_data.get("transaction_id", None)}
+        self.publish(market_stats_response_channel, json.dumps(ret_val))
+
+    def event_market_cycle(self):
+        if self.area.current_market is None:
+            return
+        market_event_channel = f"{self.channel_prefix}/market-events/market"
+        current_market_info = self.area.current_market.info
+        current_market_info['last_market_stats'] = \
+            self.area.stats.get_price_stats_current_market()
+        current_market_info["self_sufficiency"] = \
+            self.area.endpoint_stats["kpi"].get("self_sufficiency", None)
```

Comment:
What about `self-consumption` ?

Reply:
This is not requested int the ticket: https://gridsingularity.atlassian.net/browse/D3ASIM-2158